File: assignment/02/road_construction.cc
// alternative (BFS) provided by TA 鄧晉杰
template
<
  typename Graph,
  typename IsDiscovered,
  typename Costs
>
void CalculateNextIslandCost
(
  Graph const &graph,
  IsDiscovered &is_discovered,
  Costs &costs,
  long source
)
{
  costs.emplace_back(0);
  std::deque<long> nodes;
  is_discovered[source] = true;
  nodes.emplace_back(source);
  while (!nodes.empty())
  {
    auto from {nodes.front()};
    nodes.pop_front();
    for (auto to_cost : graph[from])
    {
      auto to {std::get<0>(to_cost)};
      auto cost {std::get<1>(to_cost)};
      if (!is_discovered[to])
      {
        is_discovered[to] = true;
        nodes.emplace_back(to);
      }
      costs.back() += cost;
    }
  }
  costs.back() /= 2; // hand-shaking theorem
  return;
}

std::vector<long> CalculateCosts (long n, long m, std::vector<std::vector<long>> roads)
{
  std::vector<long> costs;
  std::vector<std::vector<std::pair<long, long>>> graph(n);
  for (auto road : roads)
  {
    graph[road[0]].emplace_back(road[1], road[2]);
    graph[road[1]].emplace_back(road[0], road[2]);
  }
  std::vector<bool> is_discovered(n);
  for (long source {}; source != n; ++source)
  {
    if (!is_discovered[source])
    {
      CalculateNextIslandCost(graph, is_discovered, costs, source);
    }
  }
  std::sort(costs.begin(), costs.end());
  return costs;
}

File: assignment/02/road_construction.py
# alternative provided by TA 蔡宗霖
from collections import deque
def BFS(graph, find, start, weight):
    cost = 0
    member = 1
    queue = deque([])
    find[start] = True
    queue.append(start)
    while len(queue) > 0:
        v = queue.popleft()
        cost += weight[v]
        for nei in graph[v]:
            if not find[nei]:
                find[nei] = True
                queue.append(nei)
                member += 1
    return(cost, member)

def CalculateCosts(n, m, roads):
    adj_l = [[] for _ in range(n)]
    weight = [0] * n
    find = [False] * n

    for s, t, w in roads:
        adj_l[s].append(t)
        adj_l[t].append(s)
        weight[s] += w

    output = []
    for i in range(n):
        if not find[i]:
            cost, members = BFS(adj_l, find, i, weight)
            output.append(cost)
    return(sorted(output))
